test/opt/assembly_builder.h

```diff
@@ -0,0 +1,207 @@
+// directly. It relies on OpName instructions and friendly-name disassembling
+// to keep the ID names unchanged after assembling.
+//
+// An assembly module is divided into several sections:
+//  Global Preamble:
+//    declare capabilities, extensions, entry points, etc.
+//  Names:
+//    declare the names for IDs with OpName instructions.
+//  Types, Constants and Global variables:
+//    declare types, constants and global variables.
+//  Main Function:
+//    contains instructions in main function, just after OpLable for the entry
+//    block.
+//  Main Function Postamble:
+//    contains the return and fucntion end instruction.
+//
+// The assembly code is built by concatenating all the strings in the above
+// section.
+//
+// Users define the contents in section <Type, Constants and Global Variables>
+// and <Main Function>. The <Names> section is to hold the names for IDs to
+// keep them unchanged before and after assembling. All defined IDs to be added
+// to this code builder will be assigned with a global name through OpName
+// instruction. The name is extracted from the definition instruction.
+//  E.g. adding instruction: %var_a = OpConstant %int 2, will also add an
+//  instruction: OpName %var_a, "var_a".
+//
+// Note that the name must not be used on more than defined IDs and friendly-name
+// disassembling must be enabled so that OpName instructions will be respected.
+class AssemblyBuilder {
+ public:
+  // Initalize an minimal SPIR-V assembly code as the template. The minimal
+  // contains an empty main function and some predefined names for the main
+  // function.
+  AssemblyBuilder()
+      : global_preamble_vec_({
+            // clang-format off
+                  "OpCapability Shader",
+                  "OpCapability Float64",
+             "%1 = OpExtInstImport \"GLSL.std.450\"",
+                  "OpMemoryModel Logical GLSL450",
+                  "OpEntryPoint Vertex %main \"main\"",
+            // clang-format on
+        }),
+        names_vec_({
+            // clang-format off
+                  "OpName %main \"main\"",
+                  "OpName %main_func_entry_block \"main_func_entry_block\"",
+                  "OpName %main_func_type \"main_func_type\"",
+            // clang-format on
+        }),
+        types_consts_globals_vec_({
+            // clang-format off
+          "%void = OpTypeVoid",
+"%main_func_type = OpTypeFunction %void",
+            // clang-format on
+        }),
+        main_func_preamble_vec_({
+            // clang-format off
+          "%main = OpFunction %void None %main_func_type",
+"%main_func_entry_block = OpLabel"
+            // clang-format on
+        }),
+        main_func_entry_block_vec_(),
+        main_func_postamble_vec_({
+            // clang-format off
+                  "OpReturn",
+                  "OpFunctionEnd",
+            // clang-format on
+        }),
+        names_({{"main", 0}, {"main_func_entry_block", 1}, {"main_func_type", 2}}) {}
```

Comment:
This seems coupled to the initialization of names_vec_.   Better to leave the initialization empty, then construct the values automatically in the body of the constructor.